modules/login/php/ajaxlogin.class.inc

```diff
@@ -0,0 +1,405 @@
+            // login - user attempting to login.
+            $response = $this->login($values);
+        } else if (isset($values['command']) && $values['command'] == 'expired') {
+            // login - expired password update
+            $response = $this->expiredPassword($values);
+        } else {
+            $response['status'] = 'error';
+        }
+        return $response;
+    }
+    /**
+     * Initialize setup.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function initializeSetup(array $values) : array
+    {
+        // Login Setup.
+        $loginData = array();
+        $login     = $_SESSION['State']->getProperty('login');
+        // Previous login failed.
+        if (!empty($login->_lastError)) {
+            $loginData['error'] = $login->_lastError;
+        }
+        $config      =& \NDB_Config::singleton();
+        $study_links = $config->getExternalLinks('Studylinks');
+        foreach ($study_links as $label => $url) {
+            $windowName           = md5($url);
+            $loginData['links'][] = array(
+                                     'url'        => $url,
+                                     'label'      => $label,
+                                     'windowName' => $windowName,
+                                    );
+        }
+        $loginData['description'] = html_entity_decode(
+            $config->getSetting('StudyDescription')
+        );
+        $loginData['title']       = $config->getSetting('title');
+        $loginData['logo']        = $config->getSetting('studylogo');
+        $values['login']          = $loginData;
+
+        // Request Account Setup.
+        $requestAccountData = array();
+        $reCAPTCHAPrivate   = $config->getSetting('reCAPTCHAPrivate');
+        $reCAPTCHAPublic    = $config->getSetting('reCAPTCHAPublic');
+        // reCAPTCHA if both private and public keys are set
+        if ($reCAPTCHAPrivate && $reCAPTCHAPublic) {
+            $requestAccountData['captcha'] = $reCAPTCHAPublic; // captcha key.
+        }
+        $requestAccountData['site'] = \Utility::getSiteList();
+        $values['requestAccount']   = $requestAccountData;
+
+        return $values;
+    }
+
+    /**
+     * Reset password.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function resetPassword(array $values) : array
+    {
+        // Always claim success to phishers.
+        $values['success'] = 'Password reset. '
+            . 'You should receive an email within a few minutes.';
+
+        // Create the user object.
+        $username = $values['username'];
+        $user     =& \User::singleton($username);
+
+        // Check if valid user.
+        $email = '';
+        try {
+            $email = $user->getData('Email');
+            if (empty($email) || empty($user)) {
+                $values['error'] = 'invalid';
```

Comment:
Is this displayed to the user? We should return a more descriptive error than "invalid".

Reply:
@johnsaigle It's not displayed to the user. I reused all of the php code for the login with a few changes of my own. I'm likely not doing anything when the $values['error'] is invalid because I haven't tested a case to fire that yet.

Reply:
In that case you should provide more detailed testing instructions. At minimum make sure that the `TestPlan.md` is up-to-date with your new changes and that the module works according to what it says.

Reply:
@johnsaigle I'm hoping to see what travis complains about.

Reply:
That's fine Travis can definitely help you figure out if we can remove some manual testing steps. The TestPlan should still be up-to-date with how the module is written though.

In fact it looks like this module got passed over and has its own format for testing, which isn't good.

I'm going to tag this with `Needs Documentation`. Before this gets merged you should update consolidate the testing instructions into one document, remove typos, and make sure it accurately reflects the new state of the module. It'll help everyone with manual testing also.